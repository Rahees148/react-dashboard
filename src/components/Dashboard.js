import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import {Fragment} from 'react';

//Local-Components
import Header from './Header';
import CompletedTask from './CompletedTask';
import LatestTask from './LatestTask';
import PieChart from './PieChart';
import Tasks from './Tasks';

//Context
import {TaskContext} from '../context/TaskContext';

import NoTask from './NoTask';
import NewTask from './NewTask';



const Wrapper = styled.section`
  width: 90vw;
  margin:0 auto;
  @media (min-width: 768px) {
    width: 80vw;
  }
`;
const Gird = styled.div`
  width: 100%;
  margin-top:72px;
  padding:22px 0;
  margin-bottom:12px;
  display: grid;
  grid-gap: 8px;
  @media (min-width: 768px) {
    grid-template-columns: repeat(auto-fill,minmax(30%, 1fr));
    grid-gap: 2%;
  }
`;


function Dashboard(props){
  
    const [tasks, setTasks] = useState([]);
    const [showNewTask, setshowNewTask] = useState(false);

    useEffect(()=>{
      const taskList = localStorage.getItem('tasks');
      console.log('taskList', taskList)
      if(taskList)
      setTasks(JSON.parse(taskList));
    },[])

    const onAddItem = (item) =>{
        const newItem = [item, ...tasks]
        setTasks(newItem);
        updateLocalStorage(newItem);
    }

    const onUpdateItem = (item) =>{
      const newState = tasks.map((currentItem) => {
          if (currentItem.id === item.id) {
            return item
        } else {
          return currentItem
        }
      });
      setTasks(newState);
      updateLocalStorage( newState);
  }

    const onRemoveItem = (item) =>{   
        const remainingItems = [
          ...tasks.filter( (i) => i.id !== item.id)
        ]
        setTasks(remainingItems);
        updateLocalStorage(remainingItems);
    }

    const updateLocalStorage = (tasksList) =>{
      localStorage.setItem('tasks', JSON.stringify(tasksList));
    }

    const onToggleNewTask = () => {
        setshowNewTask( showNewTask ? false : true );
    }



    return (
        <TaskContext.Provider
        value={{
            tasks: tasks,
            addItem: onAddItem,
            removeItem: onRemoveItem,
            updateItem: onUpdateItem,
            toggleNewTask: onToggleNewTask
          }}
        >
                <Header/>
                {
                    tasks.length > 0 ? ( <Wrapper>
                    <Gird>
                        <CompletedTask />
                        <LatestTask />
                        <PieChart />
                    </Gird>
                    <Tasks />
                    </Wrapper>) : <NoTask /> }

                { showNewTask &&
                <NewTask />
                }
        </TaskContext.Provider>
       
    )
}

export default Dashboard;