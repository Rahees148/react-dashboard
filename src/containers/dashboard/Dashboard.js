/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState, useContext, Fragment } from "react";
import { trackPromise, usePromiseTracker } from "react-promise-tracker";
import { Style } from "./Dashboard.style";

//Context & service
import { apiProvider } from "../../services/provider";
import AuthApi from "../../context/AuthApi";
import { TaskContext } from "../../context/TaskContext";

//Local-Components
import Header from "../../components/header/Header";
import CompletedTask from "../../components/completedTask/CompletedTask";
import LatestTask from "../../components/latestTask/LatestTask";
import PieChart from "../../components/pieChart/PieChart";
import Tasks from "../../components/tasks/Tasks";
import NoTask from "../../components/noTask/NoTask";
import NewTask from "../../components/newTask/NewTask";

import SkeletonLoading from "../../utility/SkeletonLoding";

function Dashboard(props) {
  const { promiseInProgress } = usePromiseTracker();
  const Auth = useContext(AuthApi);
  const [tasks, setTasks] = useState([]);
  const [showNewTask, setshowNewTask] = useState(false);

  useEffect(() => {
    getAllTasks();
  }, []);

  const getAllTasks = () => {
    // const taskList = localStorage.getItem('tasks');
    trackPromise(
      apiProvider.getAll(`tasks?userId=${Auth.user.id}`).then((res) => {
        const taskList = res;
        setTasks(taskList);
      })
    );
  };

  const onAddItem = (item) => {
    const newItem = [item, ...tasks];
    setTasks(newItem);

    // updateLocalStorage(newItem);
    apiProvider.post("tasks", item);
  };

  const onUpdateItem = (item) => {
    const newState = tasks.map((currentItem) => {
      if (currentItem.id === item.id) {
        return item;
      } else {
        return currentItem;
      }
    });
    setTasks(newState);

    //updateLocalStorage( newState);
    apiProvider.put(`tasks/${item.id}`, item);
  };

  const onRemoveItem = (item) => {
    const remainingItems = [...tasks.filter((i) => i.id !== item.id)];
    setTasks(remainingItems);

    //updateLocalStorage(remainingItems);
    apiProvider.remove("tasks", item.id);
  };

  // const updateLocalStorage = (tasksList) =>{
  //   localStorage.setItem('tasks', JSON.stringify(tasksList));
  // }

  const onToggleNewTask = () => {
    setshowNewTask(showNewTask ? false : true);
  };

  return (
    <Fragment>
      <Header />
      {!promiseInProgress && (
        <TaskContext.Provider
          value={{
            tasks: tasks,
            addItem: onAddItem,
            removeItem: onRemoveItem,
            updateItem: onUpdateItem,
            toggleNewTask: onToggleNewTask,
          }}
        >
          {tasks.length > 0 ? (
            <Style.Wrapper>
              <Style.Grid>
                <CompletedTask />
                <LatestTask />
                <PieChart />
              </Style.Grid>
              <Tasks />
            </Style.Wrapper>
          ) : (
            <NoTask />
          )}
          {showNewTask && <NewTask />}
        </TaskContext.Provider>
      )}
      <SkeletonLoading />
    </Fragment>
  );
}

export default Dashboard;
