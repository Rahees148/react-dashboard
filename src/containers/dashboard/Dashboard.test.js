
import renderer from 'react-test-renderer';
import Dashboard from './Dashboard';


it('renders correctly', () => {
  const tree = renderer
    .create(<Dashboard />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});