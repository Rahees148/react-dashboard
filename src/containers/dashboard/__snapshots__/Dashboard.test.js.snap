// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <header
    className="sc-dlfnbm bAoyv"
  >
    <section
      className="sc-hKgILt gScNTu"
    >
      <div
        className="sc-eCssSg bwOWBP"
      >
        <img
          alt=""
          className="sc-jSgupP cceJut"
          src="profile.png"
        />
        
      </div>
      <a
        className="sc-gKsewC lgxjeu"
        onClick={[Function]}
      >
        Logout
      </a>
    </section>
  </header>,
  <section
    className="sc-lmoMRL kxBixP"
  >
    <div
      className="sc-iJuUWI blELXw"
    >
      <h3
        className="sc-jrAGrp dYXEGc"
      >
        You have no task.
      </h3>
      <button
        className="sc-cxFLnm gpivlI"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        + New Task
      </button>
    </div>
  </section>,
]
`;
